### PR TITLE
Adding the habit list

### LINKED ISSUE #5: Create a list of habits
**User statement**
As a doer, I want to see a list of all my habits.

**Describe the solution you'd like**
* Create a list that stores habits
* Look into implementing **RecyclerView** as its more flexible and efficient than **ListView**

**Development notes**

### PR DESCRIPTION
**Pull request by rudy-patel**
This pull request adds a list of habits and closes #5.

### Changes included
* A pre-filled RecyclerView of placeholder habits
* The design of the row is **subject to change**, this story focuses on the **implementation** of the list
* Swiping the habit to the left also **deletes** it from the list on the front-end (db integration not yet implemented)
* Clicking one of the habits produces a simple pop-up courtesy of the placeholder click listener that will later be updated

### Screenshot
Here is what the list looks like so far, please ignore the ugliness of the colors 😂
image

### CODE REVIEW

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/DashboardFragment.java [hunk 1]
```diff
@@ -1,46 +1,80 @@
...
+
+        return view;
+    }
+
+    private void initializeRecyclerView(LinearLayoutManager layoutManager, View view) {
+        RecyclerView recyclerView = view.findViewById(R.id.recycler_view_items);
+        recyclerView.setItemAnimator(new DefaultItemAnimator());
+        recyclerView.setLayoutManager(layoutManager);
+        recyclerView.setAdapter(_habitItemAdapter);
+        new ItemTouchHelper(_itemTouchHelperCallback).attachToRecyclerView(recyclerView);
```
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/DashboardFragment.java:49]: The **RecyclerView** is initialized here. The following pieces are initialized:
* Item animation
* Layout manager
* Custom adapter
* ItemTouchHelper, which adds the tap/swipe listener functionality

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitItemAdapter.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    @NonNull
+    @Override
+    public HabitViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        View view = LayoutInflater.from(_context).inflate(R.layout.habit_row, parent, false);
+        return new HabitViewHolder(view);
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull HabitViewHolder holder, int position) {
```
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitItemAdapter.java:42]: In the `onBindViewHolder`, all the front-end attributes that make up a habit row are set.

File: app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitItemAdapter.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            Toast.makeText(_context, _habits.get(position).getTitle(), Toast.LENGTH_SHORT).show();
+        });
+    }
+
+    @Override
+    public int getItemCount() {
+        return _habits.size();
+    }
+
+    public static class HabitViewHolder extends RecyclerView.ViewHolder {
```
Comment (rudy-patel) [inline app/src/main/java/com/example/habitsmasher/ui/dashboard/HabitItemAdapter.java:67]: This custom `ViewHolder` holds all the front-end attributes of the habit row in the list.

File: app/src/main/res/layout/habit_row.xml [hunk 4]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (rudy-patel) [inline app/src/main/res/layout/habit_row.xml:2]: This layout file describes the design of a single **habit row**.

File: app/src/main/res/navigation/mobile_navigation.xml [hunk 5]
```diff
@@ -14,7 +14,7 @@
     <fragment
         android:id="@+id/navigation_dashboard"
         android:name="com.example.habitsmasher.ui.dashboard.DashboardFragment"
-        android:label="@string/title_dashboard"
+        android:label="@string/habit_list"
```
Comment (rudy-patel) [inline app/src/main/res/navigation/mobile_navigation.xml:17]: This changes the menu bar string:
image

**Comment by Nova-Drive** [comment]
Is the swipe action a single swipe then delete, or is it a swipe to reveal a delete button?

**Comment by Nova-Drive** [comment]
One more thing, each time I go from the habit list to another activity like the Home Screen then come back to the habit list, it re-adds the 10 sample habits. Repeating this leads to more than the 10 sample habits. (Ex. Going back and forth 2 times adds 20 more habits to the list for a total of 30)

**Comment by rudy-patel** [comment]
> Is the swipe action a single swipe then delete, or is it a swipe to reveal a delete button?

@Nova-Drive The swipe action is a single swipe delete currently. There is a dedicated story to tackle the deletion (#8), where we will integrate this with the database. So the whole "confirm if you want to delete" workflow can be added in there! I've updated that story to include this.

File: app/src/main/res/layout/fragment_dashboard.xml [hunk 6]
```diff
@@ -2,21 +2,22 @@
...
         android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_marginStart="8dp"
-        android:layout_marginTop="8dp"
-        android:layout_marginEnd="8dp"
-        android:textAlignment="center"
-        android:textSize="20sp"
+        android:layout_height="match_parent"
+        android:clipToPadding="false"
+        android:paddingBottom="56dp"
```
Comment (rudy-patel) [inline app/src/main/res/layout/fragment_dashboard.xml:14]: This padding allows for the bottom habit to not get cut off due to the bottom navigation bar, as noticed by @Nova-Drive!

**Review by Nova-Drive — APPROVED** [review]
Works good now 👍

**Pull request merged**